test/HttpTestHelpers.h

```diff
@@ -0,0 +1,102 @@
+// Copyright 2022, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Hannah Bast (bast@cs.uni-freiburg.de)
+
+#pragma once
+
+#include <gtest/gtest.h>
+
+#include <chrono>
+#include <thread>
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_join.h"
+#include "util/http/HttpClient.h"
+#include "util/http/HttpServer.h"
+#include "util/http/HttpUtils.h"
+
+using namespace std::literals;
+
+// Test HTTP Server.
+template <typename HttpHandler>
+class TestHttpServer {
+ private:
+  // The server.
+  std::shared_ptr<HttpServer<HttpHandler>> server_;
+
+  // The own thread in which the server is running.
+  //
+  // NOTE: It is important that this thread object lives as long as the server
+  // lives. If it would be destroyed while the server is still running, the
+  // program would hang because the thread would wait for the server to exit.
+  std::unique_ptr<std::jthread> serverThread_;
+
+ public:
+  // Create server on localhost. Try out 10 different ports, if connection
+  // to all of them fail, throw `std::runtime_error`.
+  //
+  // TODO: Is there a more robust way to find a free port?
+  TestHttpServer(HttpHandler httpHandler) {
+    std::vector<short unsigned int> ports(10);
+    std::generate(ports.begin(), ports.end(),
+                  []() { return 1024 + std::rand() % (65535 - 1024); });
+    const std::string& ipAddress = "0.0.0.0";
+    int numServerThreads = 1;
+    for (const short unsigned int port : ports) {
+      try {
+        server_ = std::make_shared<HttpServer<HttpHandler>>(
+            port, ipAddress, numServerThreads, std::move(httpHandler));
+        return;
+      } catch (const boost::system::system_error& b) {
+        LOG(INFO) << "Starting test HTTP server on port " << port
+                  << " failed, trying next port ..." << std::endl;
+      }
+    }
+    throw std::runtime_error(
+        absl::StrCat("Could not start test HTTP server on any of these ports: ",
+                     absl::StrJoin(ports, ", ")));
+  }
+
+  // Get the port on which this server is running.
+  short unsigned int getPort() const { return server_->getPort(); }
+
+  // Run the server in its own thread. Wait for 100ms until the server is up and
+  // throw `std::runtime_error` if it's not (it should be up immediately).
+  //
+  // NOTE 1: It is important to *copy* the `httpServer` shared pointer into the
+  // thread. That way, whoever dies first (this thread or `httpServerThread`),
+  // the pointer is still valid in the other thread.
```

Comment:
Threads are `joined` or `run to completion`, objects `go out of scope` and then are `destroyed` via the `destructor`, none of them dies (hopefully).